pysrc/utils/sampling_methods.py

```diff
@@ -13,72 +13,106 @@ def random_sample(data: list[list[str]],
                  num_samples: int,
                  num_entries: int) -> pd.Series:
     tmp = pd.Series(data)
-    return tmp.iloc[:num_samples]
+    tmp = replaceEmptyField(tmp)
+    tmp.dropna(inplace=True)
+    tmp = tmp.loc[tmp.astype(str).drop_duplicates().index]
+
+    if len(tmp) - 1 > num_samples:
+        return tmp.iloc[:num_samples]
+    else:
+        return tmp
 
 def smallest_value_sample(data: list[list[str]],
                  num_samples: int,
                  num_entries: int) -> pd.Series:
     tmp = pd.Series(data)
     tmp = replaceEmptyField(tmp)
     tmp.dropna(inplace=True)
-    out = tmp.sort_values()
-    return out.iloc[:num_samples]
+    tmp = tmp.loc[tmp.astype(str).drop_duplicates().index]
 
+    out = tmp.sort_values()
+    if len(out) - 1 > num_samples:
+        return out.iloc[:num_samples]
+    else:
+        return out
 def biggest_value_sample(data: list[list[str]],
                  num_samples: int,
                  num_entries: int) -> pd.Series:
     tmp = pd.Series(data)
     tmp = replaceEmptyField(tmp)
     tmp.dropna(inplace=True)
-    out = tmp.sort_values(ascending=False)
-
-    return out.iloc[:num_samples]
+    tmp = tmp.loc[tmp.astype(str).drop_duplicates().index]
 
+    out = tmp.sort_values(ascending=False)
+    if len(out) - 1 > num_samples:
+        return out.iloc[:num_samples]
+    else:
+        return out
 def longest_value_sample(data: list[list[str]],
                  num_samples: int,
                  num_entries: int) -> pd.Series:
     tmp = pd.Series(data)
     tmp = replaceEmptyField(tmp)
     tmp.dropna(inplace=True)
+    tmp = tmp.loc[tmp.astype(str).drop_duplicates().index]
+
     tmp.index = tmp.str.len()
     df = tmp.sort_index(ascending=False).reset_index(drop=True)
-
-    return df.iloc[:num_samples]
+    if len(df) - 1 > num_samples:
+        return df.iloc[:num_samples]
+    else:
+        return df
 
 def shortest_value_sample(data: list[list[str]],
                  num_samples: int,
                  num_entries: int) -> pd.Series:
     tmp = pd.Series(data)
     tmp = replaceEmptyField(tmp)
     tmp.dropna(inplace=True)
+    tmp = tmp.loc[tmp.astype(str).drop_duplicates().index]
+
     tmp.index = tmp[0].str.len()
     df = tmp.sort_index(ascending=True).reset_index(drop=True)
+    if len(df) - 1 > num_samples:
+        return df.iloc[:num_samples]
+    else:
+        return df
 
-    return df.iloc[:num_samples]
-
+#Probably the same as first now and therefore obsolete?
 def all_distinct_sample(data: list[list[str]],
                  num_samples: int,
                  num_entries: int) -> pd.Series:
     tmp = pd.Series(data)
     df = tmp.loc[tmp.astype(str).drop_duplicates().index]
-
-
     return df.iloc[:num_samples]
 
+#only method were I had to make it dependent on sample size after deduplicating
+#no idea how to keep it there because at some points it tries to sample from spaces that doesn't exist
+#doesnt't work?
```

Comment:
Maybe one or both `ceil` should be `floor`? Or instead of `i % num_entries` use `i % len(tmp)`?

Reply:
I tried a bit around i % len(tmp) would work but somehow in the method i am loosing a factor 10 of values, but not entirely sure why tho so far.